spec.md

```diff
@@ -2734,6 +2780,270 @@ message NodeExpandVolumeResponse {
+
+If a group snapshot corresponding to the specified group snapshot `name` is successfully created (meaning all snapshots associated with the group are successfully cut), the Plugin MUST reply `0 OK` with the corresponding `CreateVolumeGroupSnapshotResponse`.
+
+If an error occurs before a group snapshot is cut, `CreateVolumeGroupSnapshot` SHOULD return a corresponding gRPC error code that reflects the error condition.
+
+`CreateVolumeGroupSnapshot` SHOULD return `0 OK` after all the snapshots are cut regardless of whether post processing such as uploading is complete or not.
+
+CO SHOULD check the `ready_to_use` boolean of individual snapshots that are members of the group snapshot and only treat the group snapshot as ready to use when all snapshots have been "processed" and is ready to use to create new volumes from those snapshots.
+
+An individual snapshot MAY be used as the source to provision a new volume.
+
+In VolumeGroupSnapshot message, both snapshots and group_snapshot_id are required fields.
+
+If an error occurs before all the individual snapshots are created when creating a group snapshot of multiple volumes and a group_snapshot_id is not yet available for CO to do clean up, SP SHOULD do clean up and make sure no snapshots are leaked.
+
+```protobuf
+message CreateVolumeGroupSnapshotRequest {
+  option (alpha_message) = true;
+
+  // The suggested name for the group snapshot. This field is REQUIRED
+  // for idempotency.
+  // Any Unicode string that conforms to the length limit is allowed
+  // except those containing the following banned characters:
+  // U+0000-U+0008, U+000B, U+000C, U+000E-U+001F, U+007F-U+009F.
+  // (These are control characters other than commonly used whitespace.)
+  string name = 1;
+
+  // volume ids of the source volumes to be snapshotted together.
+  // This field is REQUIRED.
+  repeated string source_volume_ids = 2;
+
+  // Secrets required by plugin to complete
+  // ControllerCreateVolumeGroupSnapshot request.
+  // This field is OPTIONAL. Refer to the `Secrets Requirements`
+  // section on how to use this field.
+  // The secrets provided in this field SHOULD be the same as
+  // the secrets provided in ControllerDeleteVolumeGroupSnapshot
+  // and ControllerGetVolumeGroupSnapshot requests for the same
+  // group snapshot unless if secrets are rotated after the
+  // group snapshot is created.
+  map<string, string> secrets = 3 [(csi_secret) = true];
+
+  // Volume secrets required by plugin to complete volume group
+  // snapshot creation request. This field is needed in case the
+  // volume level secrets are different from the above secrets
+  // for the group snapshot.
+  // This field is OPTIONAL.
+  repeated VolumeSecret volume_secrets = 4;
```

Comment:
are there SPs that actually need this for group snapshotting? if not can we table this for later? we have secrets APIs bleeding all over the spec, and if no system needs this right now then why add it?

Reply:
I don't have a concrete example of a SP needing this.  @humblec told me Ceph does not need this as Secrets and VolumeSecrets will be the same.  This is added just in case some SP might need this to support per-volume secrets.

@bswartz Are you okay if we drop this field for now?

Reply:
Checked with @ameade from NetApp and @ggriffiths from Portworx.  They don't need VolumeSecrets for the group snapshots.  So dropping it for now.